### ISSUE TITLE
[Quant] [PT2E] PT2 QAT Quantization gives worse accuracy than FX QAT Quantization 

### ISSUE THREAD

**Issue by leslie-fang-intel** [pytorch/pytorch#111192]
### 🐛 Describe the bug

Compare RN50 QAT Quantization accuracy between:

- PT2 QAT Quantization with `XNNPACKQuantizer`. 
  - E2E Accuracy is: `* Acc@1 57.216 Acc@5 92.032`.
  - Test Script is: https://gist.github.com/leslie-fang-intel/bc606f29ea927e8141c9d9efd8e8bd69
- FX QAT Quantization with `X86` Backend. 
  - E2E Accuracy is: ` * Acc@1 75.498 Acc@5 92.626`.
  - Test Script is: https://gist.github.com/leslie-fang-intel/e5d1496cf5ec4b84f85c1248e0296ddf

Looks like the accuracy of PT2 QAT Quantization with `XNNPACKQuantizer` is much worse than FX QAT Quantization with `X86` Backend.

### Versions
```
(quantization) [root@CPX-4 quantization]# python collect_env.py
Collecting environment information...
PyTorch version: 2.2.0a0+gitc679db2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 11.2.1 20210728 (Red Hat 11.2.1-1)
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.17

Python version: 3.9.10 (main, Mar  2 2022, 12:02:00)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-4.19.5-1.el7.elrepo.x86_64-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                56
On-line CPU(s) list:   0-55
Thread(s) per core:    1
Core(s) per socket:    28
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel Genuine CPU
Stepping:              10
CPU MHz:               2899.109
CPU max MHz:           2900.0000
CPU min MHz:           1200.0000
BogoMIPS:              5800.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              39424K
NUMA node0 CPU(s):     0-27
NUMA node1 CPU(s):     28-55
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] flake8==3.8.2
[pip3] flake8-bugbear==20.1.4
[pip3] flake8-coding==1.3.3
[pip3] flake8-comprehensions==3.3.0
[pip3] flake8-executable==2.0.4
[pip3] flake8-pyi==20.5.0
[pip3] intel-extension-for-pytorch==2.2.0+git2341961
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.22.3
[pip3] torch==2.2.0a0+git5614023
[pip3] torchvision==0.16.0a0+370134
[pip3] triton==2.0.0
[conda] mkl                       2023.0.0            intel_25398    intel
[conda] mkl-include               2023.0.0            intel_25398    intel
[conda] mkl-include               2023.0.0                  <pip>
[conda] mkl-service               2.4.0           py39h3605609_14    intel
[conda] mkl-static                2023.0.0                  <pip>
[conda] mkl_fft                   1.3.1           py39hcab1719_22    intel
[conda] mkl_random                1.2.2           py39hbf47bc3_22    intel
[conda] mkl_umath                 0.1.1           py39hf66a691_32    intel
[conda] numpy                     1.22.3           py39hf0956d0_5    intel
[conda] numpy-base                1.22.3           py39h45c9ace_5    intel
[conda] torchvision               0.16.0a0+370134           <pip>
[conda] triton                    2.0.0                     <pip>
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen

**Comment by leslie-fang-intel** [comment]
Hi @andrewor14 @jerryzh168, could you kindly help to take a look of this issue?

**Comment by andrewor14** [comment]
Hi @leslie-fang-intel, this is a known issue for models with batch norm. Can you try following the "Note" under [this section of the QAT tutorial](https://pytorch.org/tutorials/prototype/pt2e_quant_qat.html#prepare-the-model-for-quantization-aware-training)? The high level problem is the default CPU batch norm op is known to have worse numerics than the cuda batch norm op, so you may need to add some manual op swapping logic.

**Comment by leslie-fang-intel** [comment]
Hi @andrewor14, thanks for the feedback. 

- May I know is there any issue ticket to track it? 
_The high level problem is the default CPU batch norm op is known to have worse numerics than the cuda batch norm op_

- Will take a try of the Note in [this section of the QAT tutorial](https://pytorch.org/tutorials/prototype/pt2e_quant_qat.html#prepare-the-model-for-quantization-aware-training) to set up the CUDA ENV.

**Comment by andrewor14** [comment]
Here's the issue: https://github.com/pytorch/pytorch/issues/111384

**Comment by leslie-fang-intel** [comment]
Thanks a lot @andrewor14. Tried your suggestions:

- Step 1: QAT Training

  - Load model and to CUDA
  - Prepare Model with X86InductorQuantizer
  - QAT Training Loop (only do 2 steps as a example)
  - Save the checkpoint

- Step 2: Convert and Lowering

  - Load model and to CPU
  - Prepare Model with X86InductorQuantizer
  - Load the QAT checkpoint with weights generated in step 1
  - Convert PT2E
  - `Torch.compile` lowers to Inductor
  - Validation Loop

I can get E2E accuracy of `* Acc@1 76.026 Acc@5 92.872` now which is close to PTQ.
I think we can close this ticket and track the BN numeric issue in https://github.com/pytorch/pytorch/issues/111384

**Comment by andrewor14** [comment]
Ok sounds good. Thanks!

**Comment by jerryzh168** [comment]
> Thanks a lot @andrewor14. Tried your suggestions:
> 
> * Step 1: QAT Training
>   
>   * Load model and to CUDA
>   * Prepare Model with X86InductorQuantizer
>   * QAT Training Loop (only do 2 steps as a example)
>   * Save the checkpoint
> * Step 2: Convert and Lowering
>   
>   * Load model and to CPU
>   * Prepare Model with X86InductorQuantizer
>   * Load the QAT checkpoint with weights generated in step 1
>   * Convert PT2E
>   * `Torch.compile` lowers to Inductor
>   * Validation Loop
> 
> I can get E2E accuracy of `* Acc@1 76.026 Acc@5 92.872` now which is close to PTQ. I think we can close this ticket and track the BN numeric issue in #111384

@leslie-fang-intel what did you change in https://gist.github.com/leslie-fang-intel/bfc077f040f605c2102acea07e5b23c1? the example code is not doing training as mentioned by @andrewor14

**Comment by leslie-fang-intel** [comment]
Hi @jerryzh168, sorry, I forget to update the script.

- Here is the QAT training script https://github.com/leslie-fang-intel/torch_script/blob/master/inductor/int8/qat/x86inductorquantizer_qat_acc_cuda.py corresponding to above step 1. 
  - Step 1: QAT Training
    - Load model and to CUDA
    - Prepare Model with X86InductorQuantizer
    - QAT Training Loop (only do 2 steps as a example)
    - Save the checkpoint

- Here is the inference step script https://github.com/leslie-fang-intel/torch_script/blob/master/inductor/int8/qat/x86inductorquantizer_qat_acc_with_checkpoint.py which corresponding to above step 2.
  - Step 2: Convert and Lowering
    - Load model and to CPU
    - Prepare Model with X86InductorQuantizer
    - Load the QAT checkpoint with weights generated in step 1
    - Convert PT2E
    - Torch.compile lowers to Inductor
    - Validation Loop

**Comment by jerryzh168** [comment]
> Hi @jerryzh168, sorry, I forget to update the script.
> 
> * Here is the QAT training script https://github.com/leslie-fang-intel/torch_script/blob/master/inductor/int8/qat/x86inductorquantizer_qat_acc_cuda.py corresponding to above step 1.
>   
>   * Step 1: QAT Training
>     
>     * Load model and to CUDA
>     * Prepare Model with X86InductorQuantizer
>     * QAT Training Loop (only do 2 steps as a example)
>     * Save the checkpoint
> * Here is the inference step script https://github.com/leslie-fang-intel/torch_script/blob/master/inductor/int8/qat/x86inductorquantizer_qat_acc_with_checkpoint.py which corresponding to above step 2.
>   
>   * Step 2: Convert and Lowering
>     
>     * Load model and to CPU
>     * Prepare Model with X86InductorQuantizer
>     * Load the QAT checkpoint with weights generated in step 1
>     * Convert PT2E
>     * Torch.compile lowers to Inductor
>     * Validation Loop

OK thanks, is it fair to say that the flow doesn't work before because it is not doing training?

**Comment by leslie-fang-intel** [comment]
Yean.

- When switch to actually doing training, the accuracy in my example looks good now for PT2 QAT Path.
- One left question is why FX Quantization QAT path accuracy looks good even we are not actually doing training, for example https://gist.github.com/leslie-fang-intel/e5d1496cf5ec4b84f85c1248e0296ddf. Any idea about the difference?
